controllers/alert_controller.go

```diff
@@ -99,95 +100,80 @@ func (r *AlertReconciler) SetupWithManagerAndOptions(mgr ctrl.Manager, opts Aler
+		// Record Prometheus metrics.
+		r.Metrics.RecordReadiness(ctx, obj)
+		r.Metrics.RecordDuration(ctx, obj, reconcileStart)
+		r.Metrics.RecordSuspend(ctx, obj, obj.Spec.Suspend)
 
-			readyCondition := conditions.Get(alert, meta.ReadyCondition)
-			switch readyCondition.Status {
-			case metav1.ConditionFalse:
-				// As we are no longer reconciling and the end-state is not ready, the reconciliation has stalled
-				conditions.MarkStalled(alert, readyCondition.Reason, readyCondition.Message)
-			case metav1.ConditionTrue:
-				// As we are no longer reconciling and the end-state is ready, the reconciliation is no longer stalled
-				conditions.Delete(alert, meta.StalledCondition)
-			}
+		// Emit warning event if the reconciliation failed.
+		if retErr != nil {
+			r.Event(obj, corev1.EventTypeWarning, meta.FailedReason, retErr.Error())
 		}
 
-		if err := patchHelper.Patch(ctx, alert, patchOpts...); err != nil {
-			retErr = kerrors.NewAggregate([]error{retErr, err})
+		// Log and emit success event.
+		if retErr == nil && conditions.IsReady(obj) {
+			msg := fmt.Sprintf("Reconciliation finished in %s",
+				time.Since(reconcileStart).String())
+			log.Info(msg)
+			r.Event(obj, corev1.EventTypeNormal, meta.SucceededReason, msg)
 		}
-
-		r.Metrics.RecordReadiness(ctx, alert)
-		r.Metrics.RecordDuration(ctx, alert, start)
 	}()
 
-	if !controllerutil.ContainsFinalizer(alert, v1beta1.NotificationFinalizer) {
-		controllerutil.AddFinalizer(alert, v1beta1.NotificationFinalizer)
+	if !controllerutil.ContainsFinalizer(obj, apiv1.NotificationFinalizer) {
+		controllerutil.AddFinalizer(obj, apiv1.NotificationFinalizer)
 		result = ctrl.Result{Requeue: true}
 		return
 	}
 
-	if !alert.ObjectMeta.DeletionTimestamp.IsZero() {
-		controllerutil.RemoveFinalizer(alert, v1beta1.NotificationFinalizer)
+	if !obj.ObjectMeta.DeletionTimestamp.IsZero() {
+		controllerutil.RemoveFinalizer(obj, apiv1.NotificationFinalizer)
 		result = ctrl.Result{}
 		return
 	}
 
-	return r.reconcile(ctx, alert)
+	// Return early if the object is suspended.
+	if obj.Spec.Suspend {
+		log.Info("Reconciliation is suspended for this object")
+		return ctrl.Result{}, nil
+	}
+
+	return r.reconcile(ctx, obj)
 }
 
-func (r *AlertReconciler) reconcile(ctx context.Context, alert *v1beta1.Alert) (ctrl.Result, error) {
+func (r *AlertReconciler) reconcile(ctx context.Context, alert *apiv1.Alert) (ctrl.Result, error) {
 	// Mark the resource as under reconciliation
-	conditions.MarkReconciling(alert, meta.ProgressingReason, "")
+	conditions.MarkReconciling(alert, meta.ProgressingReason, "Reconciliation in progress")
 
 	// validate alert spec and provider
 	if err := r.validate(ctx, alert); err != nil {
-		conditions.MarkFalse(alert, meta.ReadyCondition, v1beta1.ValidationFailedReason, err.Error())
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+		conditions.MarkFalse(alert, meta.ReadyCondition, apiv1.ValidationFailedReason, err.Error())
```

Comment:
When the Provider is not found, due to the above, the following condition is observed:

```yaml
  - lastTransitionTime: "2022-11-14T14:05:22Z"
    message: 'failed to get provider ''default/slack-test2'': Provider.notification.toolkit.fluxcd.io
      "slack-test2" not found'
    observedGeneration: 2
    reason: ValidationFailed
    status: "False"
    type: Ready
```

Marking this with reason `ValidationFailed` doesn't sound right. It's misleading. The Alert configuration is valid. When the Provider is created later, the same Alert will become Ready.

In case of Receiver, when the Secret isn't found, it fails with the reason `TokenNotFound`, which is slightly better. It's clear that something outside the CRD has a problem.

`validate()` isn't really doing validation but checking if the Provider is ready. Maybe the function should be renamed to something more appropriate.